haystack/retriever/dpr_utils.py

```diff
@@ -0,0 +1,408 @@
+#!/usr/bin/env python3
+# Utilility functions and classes required for DensePassageRetriever
+#
+# Building upon the code (https://github.com/facebookresearch/DPR) published by Facebook research under Creative Commons License (https://github.com/facebookresearch/DPR/blob/master/LICENSE)
+# It is based on the following research work:
+# Karpukhin, Vladimir, et al. "Dense Passage Retrieval for Open-Domain Question Answering." arXiv preprint arXiv:2004.04906 (2020).
+# (https://arxiv.org/abs/2004.04906)
+
+import logging
+from typing import Tuple, Union, List
+
+import gzip
+import os
+import pathlib
+import wget
+
+import torch
+from torch import nn
+from torch import Tensor as T
+from torch.serialization import default_restore_location
+import collections
+from farm.file_utils import http_get
+
+from transformers.tokenization_bert import BertTokenizer
+from transformers.modeling_bert import BertModel, BertConfig
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+CheckpointState = collections.namedtuple("CheckpointState",
+                                         ['model_dict', 'optimizer_dict', 'scheduler_dict', 'offset', 'epoch',
+                                          'encoder_params'])
+
+
+# CLASSES
+class HFBertEncoder(BertModel):
+
+    def __init__(self, config, project_dim: int = 0):
+        BertModel.__init__(self, config)
+        assert config.hidden_size > 0, 'Encoder hidden_size can\'t be zero'
+        self.encode_proj = nn.Linear(config.hidden_size, project_dim) if project_dim != 0 else None
+        self.init_weights()
+
+    @classmethod
+    def init_encoder(cls, cfg_name: str, projection_dim: int = 0, dropout: float = 0.1, **kwargs) -> BertModel:
+        cfg = BertConfig.from_pretrained(cfg_name if cfg_name else 'bert-base-uncased')
+        if dropout != 0:
+            cfg.attention_probs_dropout_prob = dropout
+            cfg.hidden_dropout_prob = dropout
+        return cls.from_pretrained(cfg_name, config=cfg, project_dim=projection_dim, **kwargs)
+
+    def forward(self, input_ids: torch.Tensor, token_type_ids: torch.Tensor, attention_mask: torch.Tensor) -> Tuple[torch.Tensor, ...]:
+        if self.config.output_hidden_states:
+            sequence_output, pooled_output, hidden_states = super().forward(input_ids=input_ids,
+                                                                            token_type_ids=token_type_ids,
+                                                                            attention_mask=attention_mask)
+        else:
+            hidden_states = None
+            sequence_output, pooled_output = super().forward(input_ids=input_ids, token_type_ids=token_type_ids,
+                                                             attention_mask=attention_mask)
+
+        pooled_output = sequence_output[:, 0, :]
```

Comment:
I know this line is also in [FB repo](https://github.com/facebookresearch/DPR/blob/master/dpr/models/hf_models.py#L127), but does it make sense?

I think it only makes sense for architectures that have a CLS token at that position, otherwise I would rely on the pooled output directly. This is of course for Bert, so it can be assured that there is the CLS token at index 0.

Reply:
Agree. As you only use BERT models anyway right now it's not a problem, but I will add a comment here to clarify that this has to change if we extend it to other models.